### DISCUSSION TITLE
LinearMouse will change its code signing certificate after v0.6.1

### DISCUSSION THREAD

**Discussion by lujjjh** [linearmouse/linearmouse#141] [Announcements]
**TL;DR: I will sign LinearMouse with a new, long-term and trusted certificate in the future versions. You will need to [re-grant Accessibility permission](https://github.com/linearmouse/linearmouse/blob/main/ACCESSIBILITY.md#upgrade-from-v061-or-earlier-versions) by removing and re-add LinearMouse in Accessibility permissions after the upgrade.**

https://user-images.githubusercontent.com/3000535/173173057-ac4c8590-a4fa-48af-8559-dc9953798ce7.mp4

## Background

macOS' Gatekeeper prevents users from running apps downloaded from unidentified developers by default. That's why I had to write a detailed instruction on [how to install LinearMouse](https://github.com/linearmouse/linearmouse/blob/be363941980db763d44a5a44dfe73a26b64cfbac/README.md#manually).

The only way to become an “identified developer” is to pay Apple US$99/yr and enroll in the Apple Developer Program. LinearMouse is a free and open-source app, so I wasn't willing to do so.

So I was always using a Apple Development certificate to sign LinearMouse.

## Problems of development certificates

Although a development certificate seems to work, there are still some problems.

In addition to being a pain to install, signing apps with a development certificate is also unstable. The development certificate is only valid for one year. I must also be very careful in case of [revoking the development certificate by accident](https://github.com/linearmouse/linearmouse/issues/43#issuecomment-1003004118).

## Switching to a new certificate

Thanks to all of the sponsors, I received $76 on [Ko-fi](https://ko-fi.com/lujjjh) and $24.39 on [Open Collective](https://opencollective.com/linearmouse).

I decided to enroll in the Apple Developer Program and switch to a new certificate to make LinearMouse easier to use and less bothersome to you.

To switch to the new certificate as painless as possible:

1. I removed the logic that checks for certificate matching in v0.6.1, allowing you to upgrade to a newer version signed by a newer certificate.
2. Versions below v0.6.1 can only be directly upgraded to v0.6.1, and then to a newer version.

## What should I do?

You will need to re-grant Accessibility permission after upgrading to a new version. The instruction can be found [here](https://github.com/linearmouse/linearmouse/blob/main/ACCESSIBILITY.md#upgrade-from-v061-or-earlier-versions).

Thank you for supporting LinearMouse!

**Comment by AndreBaumeier** [comment]
Can you enable recurring contributions on Open Collective? that 99 USD will have to be payed again next year and I'd like to chip in ;)

**Reply by lujjjh** [reply]
Sure! I've enabled recurring contributions. Thank you so much for your support!

**Comment by stdc105** [comment]
Thanks for the great work!

The upgrade is pretty smooth. The prompt for re-granting Accessibility permission, however, is kinda dangerous to have a button to reset _all_ permissions without confirmation. I took it for resetting all permissions about LinearMouse, but it reset everything in Accessibility and froze my OS. Perhaps a warning dialog that explicitly states about the full reset would help here?

**Reply by lujjjh** [reply]
How about “By doing this, the Accessibility permissions will be completely reset and all permissions—including those granted to other apps—will be removed.”?

**Reply by stdc105** [reply]
That would be very helpful! Just to confirm, is this only necessary under rare circumstances that the permissions are broken, and toggling Accessibility on and off for LinearMouse does not work? If that's the case, it might be better to say that one doesn't _have_ to press this button, unless it's absolutely not working.

**Reply by lujjjh** [reply]
Yes, you are indeed correct. This button is designed for those who are not able to fix the broken permissions caused by the certificate change.

Since the certificate is not about to change for a long time, I'm going to remove it or hide it in some way in the next major version.

**Reply by lujjjh** [reply]
<img width="372" alt="image" src="https://user-images.githubusercontent.com/3000535/176609322-4316aef0-f7ae-44e8-b05a-2030f1c6d03a.png">
